Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm

```diff
@@ -5746,16 +5742,44 @@ static VisiblePositionRange constrainRangeToSelection(const VisiblePositionRange
         }();
 
         if (!scrollableArea)
-            continue;
+            return { };
 
         auto scrollingNodeID = scrollableArea->scrollingNodeID();
         if (!scrollingNodeID)
-            continue;
+            return { };
 
-        state.visualData->enclosingScrollPosition = scrollableArea->scrollPosition();
-        state.visualData->enclosingScrollingNodeID = WTFMove(scrollingNodeID);
-        break;
+        return { scrollableArea->scrollPosition(), WTFMove(scrollingNodeID) };
+    };
+
+    CheckedPtr<RenderLayer> scrollableLayer;
+    for (CheckedPtr layer = nextScroller(*enclosingLayer, IncludeSelfOrNot::IncludeSelf); layer; layer = nextScroller(*layer, IncludeSelfOrNot::ExcludeSelf)) {
+        if (auto [scrollPosition, scrollingNodeID] = scrollPositionAndNodeIDForLayer(layer.get()); scrollingNodeID) {
+            state.visualData->enclosingScrollPosition = scrollPosition;
+            state.visualData->enclosingScrollingNodeID = WTFMove(scrollingNodeID);
+            scrollableLayer = WTFMove(layer);
+            break;
+        }
+    }
+
+    if (!state.visualData->enclosingScrollingNodeID)
+        return;
+
+    if (selection.isCaret()) {
+        state.visualData->scrollingNodeIDAtStart = state.visualData->enclosingScrollingNodeID;
+        state.visualData->scrollingNodeIDAtEnd = state.visualData->enclosingScrollingNodeID;
+        return;
     }
+
+    auto scrollingNodeIDForEndpoint = [&](RenderLayer* endpointLayer) -> std::optional<ScrollingNodeID> {
```

Comment:
Is this trailing return type needed? Since both `return` statements return exactly `std::optional<ScrollingNodeID>`

Reply:
Yep, that works! Changed to omit trailing return type.